src/libostree/ostree-repo-libarchive.c

```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
-  g_autofree char *hardlink_source_checksum = NULL;
-  glnx_unref_object OstreeMutableTree *hardlink_source_subdir = NULL;
-  g_autofree guchar *tmp_csum = NULL;
-  g_autofree char *tmp_checksum = NULL;
+  g_autoptr(GPtrArray) components = NULL;
+  if (!ot_util_path_split_validate (path, &components, error))
+    return FALSE;
 
-  pathname = archive_entry_pathname (entry); 
-      
-  if (!ot_util_path_split_validate (pathname, &split_path, error))
-    goto out;
+  if (components->len == 0) /* root dir? */
+    {
+      *out_dir = g_object_ref (ctx->root);
+      return TRUE;
+    }
+
+  if (ostree_mutable_tree_walk (ctx->root, components, 0, out_dir, error))
+    return TRUE; /* already exists, nice! */
 
-  if (split_path->len == 0)
+  if (!g_error_matches (*error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+    return FALSE; /* some other error occurred */
+
+  if (ctx->opts->autocreate_parents)
     {
-      parent = NULL;
-      basename = NULL;
+      g_clear_error (error);
+      return aic_create_parent_dirs (ctx, components, out_dir,
+                                     cancellable, error);
     }
-  else
+
+  return FALSE;
+}
+
+static gboolean
+aic_get_xattrs (OstreeRepoArchiveImportContext *ctx,
+                const char         *path,
+                GFileInfo          *file_info,
+                GVariant          **out_xattrs,
+                GCancellable       *cancellable,
+                GError            **error)
+{
+  g_autofree char *abspath = g_build_filename ("/", path, NULL);
+  g_autoptr(GVariant) xattrs = NULL;
+
+  if (ctx->modifier && ctx->modifier->xattr_callback)
+    xattrs = ctx->modifier->xattr_callback (ctx->repo, abspath, file_info,
+                                            ctx->modifier->xattr_user_data);
+
+  if (ctx->modifier && ctx->modifier->sepolicy)
     {
-      if (tmp_dir_csum)
-        {
-          g_free (tmp_checksum);
-          tmp_checksum = ostree_checksum_from_bytes (tmp_dir_csum);
-          if (!ostree_mutable_tree_ensure_parent_dirs (root, split_path,
-                                                       tmp_checksum,
-                                                       &parent,
-                                                       error))
-            goto out;
-        }
-      else
-        {
-          if (!ostree_mutable_tree_walk (root, split_path, 0, &parent, error))
-            goto out;
-        }
-      basename = (char*)split_path->pdata[split_path->len-1];
+      mode_t mode = g_file_info_get_attribute_uint32 (file_info, "unix::mode");
+      g_autoptr(GVariantBuilder) builder =
+        ot_util_variant_builder_from_variant (xattrs, G_VARIANT_TYPE
+                                                        ("a(ayay)"));
+
+      if (!builder_add_label (builder, ctx->modifier->sepolicy, abspath, mode,
+                              cancellable, error))
+        return FALSE;
+
+      if (xattrs)
+        g_variant_unref (xattrs);
```

Comment:
Hmm, I don't think so. If `builder_add_label()` fails, the xattrs will be cleaned up by the cleanup func. If `builder_add_label()` succeeds, we need to drop the ref we have to the user's xattrs, if any, so that we can replace it with the variant ref we get from our builder (which already has the xattrs the user returned + the selabel).